### PR TITLE
Add support for environment variables with default values in fig.yml

### LINKED ISSUE #495: Pass variables inside fig.yml on runtime
Is it possible to configure a fig.yml in such way that you can pass on execution time variables instead of hard-wiring them inside the fig.yml?

I think a generic variable injection during execution would be quite useful for various use cases.

e.g:

<pre>
jenkins:
  image: aespinosa/jenkins:latest
  ports:
    - "8080"
  hostname: ${HOSTNAME}
</pre>

HOSTNAME=ci fig up

That could inject the variable HOSTNAME inside the fig.yml during execution and execute a docker run with hostname `ci`.

ps. This is different than passing environment variables inside docker which is already supported (http://www.fig.sh/yml.html#environment)

### PR DESCRIPTION
**Pull request by relwell**
This change resolves #495 by introducing a Config class inside the cli
library. The Config class is responsible for loading a .yml file and
evaluating its contents for environment variables. The Config immediately
reloads the data as a dictionary, which reduces compatibility issues within
previously written code. This behavior also prevents race conditions
caused by changes to the environment during long-running fig builds.

This change is valuable because it allows for a greater degree of
automation and variability when working with multiple environments
or deployment types that use identically built containers. We
currently have to define separate services for each degree of
variability we want to account for.

Future steps for this change may include introducing template files,
or adding access and manipulation APIs directly so that we can
perpetuate the config class everywhere we are currently using
bare dictionaries.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by: Robert Elwell <robert.elwell@gmail.com>
Sponsored by: Lookout, Inc.

### CODE REVIEW

File: docs/yml.md [hunk 1]
```diff
@@ -10,6 +10,14 @@ Each service defined in `fig.yml` must specify exactly one of `image` or `build`
 
 As with `docker run`, options specified in the Dockerfile (e.g. `CMD`, `EXPOSE`, `VOLUME`, `ENV`) are respected by default - you don't need to specify them again in `fig.yml`.
 
+Fig supports using environment variables with optional backoffs as values. The accepted patterns are as follows:
```
Comment (dnephin) [inline docs/yml.md]: Minor terminology here, I would expect "default" here instead of "backoff"

File: fig/cli/config.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import six
+import errno
+import yaml
+import os
+import re
+
+from . import errors
+
+
+class Config:
```
Comment (dnephin) [inline fig/cli/config.py]: without an `object` base these are old style classes

```
class Config(object):
   ...
```

File: fig/cli/config.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    def fully_interpolated(cls, value):
+        """
+        Matches our environment variable pattern, replaces the value with the value in the env
+
+        A future improvement here would be to also reference a separate dictionary that keeps track of
+        configurable variables via flat file.
+
+        :param value: any value that is reachable by a key in a dict represented by a yaml file
+        :return: the value itself if not a string with an environment variable, otherwise the value specified in the env
+        """
```
Comment (dnephin) [inline fig/cli/config.py]: This function doesn't reference `cls` at all. Could we move it off the class?

To bikeshed the name a bit, how about `resolve_environment_vars()` or something like that? Right now it reads more like a property that returns a boolean value.

File: fig/cli/config.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        if not (isinstance(value, str) or type(value) == unicode):
+            return value
+
+        env_regex = r'\$\{(?P<env_var>[^\}^:]+)(:(?P<default_val>[^\}]+))?\}'
```
Comment (dnephin) [inline fig/cli/config.py]: I worry about regexs being error prone. 

If we used `{VAR:default}` (with no `$` prefix), I suspect we could subclass https://docs.python.org/2/library/string.html#string-formatting and support this pretty easily

File: docs/yml.md (hunk 1 above)
Comment (relwell) [inline docs/yml.md]: Noted, will fix

File: fig/cli/config.py (hunk 2 above)
Comment (relwell) [inline fig/cli/config.py]: Okay, changing

File: fig/cli/config.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        """
+        Matches our environment variable pattern, replaces the value with the value in the env
+
+        A future improvement here would be to also reference a separate dictionary that keeps track of
+        configurable variables via flat file.
+
+        :param value: any value that is reachable by a key in a dict represented by a yaml file
+        :return: the value itself if not a string with an environment variable, otherwise the value specified in the env
+        """
+        if not (isinstance(value, str) or type(value) == unicode):
```
Comment (dnephin) [inline fig/cli/config.py]:
```
import six

if not isinstance(value, six.string_types):
    ...
```

py3 isn't a hard requirement yet, but it's always nice support it where it's easy

File: fig/cli/config.py (hunk 3 above)
Comment (relwell) [inline fig/cli/config.py]: The idea was to make it a classmethod since it didn't actually rely on any instance variables. What I can do is move it out of the class invoke it from the module.

File: fig/cli/config.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+    def as_dict(self):
+        """
+        Recursively forces a preprocessed config back into dict type
+        This is much easier than rewriting all client code to use a different class,
+        while providing us with a optional upgrade path (implementing Config directly where possible)
+
+        :return: the config as represented in a dict format
+        :rtype: dict
+        """
```
Comment (dnephin) [inline fig/cli/config.py]: It seems like we're having to recurse over the entire config twice. Once in the constructor, and again in `as_dict()`. 

I understand the desire for a `Config` object, but I think to really gain much over using a dict, we'd need a well defined object like a `namedtuple`.

I like that all the changes in this PR are isolated, I definitely this we can wait on the well defined `Config` object.

So instead of this double recurse, maybe this could just be a function instead of a class? It would look a lot like as_dict() without the first clause checking for `Config` objects.

File: fig/cli/config.py (hunk 3 above)
Comment (dnephin) [inline fig/cli/config.py]: that'd be great, thanks!

**Comment by dnephin** [comment]
Thanks for the PR!

I like the direction this is taking, I've left a few comments. I suspect others will be interested in this as well.

File: fig/cli/config.py (hunk 4 above)
Comment (relwell) [inline fig/cli/config.py]: Let me put some thought into this one. I think that the dollar sign is a good indicator that the information is expected to come from the environment. The colon also hints at the interpolated format of the passed value in the base version of this class, so I'm concerned there may be some problematic side effects to validate against here.

File: fig/cli/config.py (hunk 6 above)
Comment (relwell) [inline fig/cli/config.py]: I can get behind that, especially since this is a pretty isolated change.

File: fig/cli/config.py (hunk 4 above)
Comment (dnephin) [inline fig/cli/config.py]: Cool, sounds good. I want to look into this myself as well. It might even be possible to have this `Formatter` support the `$` prefix, not sure yet

**Comment by relwell** [comment]
Glad to hear it! Do you prefer an additional commit or an amended commit for revisions?

File: fig/cli/config.py (hunk 4 above)
Comment (relwell) [inline fig/cli/config.py]: I'm going to explore extending [string.Template](https://docs.python.org/2/library/string.html#string.Template), and seeing if I can change substitute() to include accessing the environment

Comment (relwell) [inline fig/cli/config.py]: It appears that using string.Template would also require regular expressions, so it wouldn't be much better than what's proposed. And string.Formatter appears to bury most of its logic in str and unicode built-ins.

I know that regular expressions can be tricky, but they are more performant and -- when used sparingly -- can be easier to maintain than creating a full-scale parser for this use case. 

Let me know if there's something I can do to help make you feel a bit better about this implementation detail.

File: docs/yml.md [hunk 7]
```diff
@@ -10,6 +10,14 @@ Each service defined in `fig.yml` must specify exactly one of `image` or `build`
 
 As with `docker run`, options specified in the Dockerfile (e.g. `CMD`, `EXPOSE`, `VOLUME`, `ENV`) are respected by default - you don't need to specify them again in `fig.yml`.
 
+Fig supports using environment variables with optional defaults as values. The accepted patterns are as follows:
+
+```
+${ENVIRONMENT_VARIABLE:backoff}
```
Comment (thaJeztah) [inline docs/yml.md]: s/backoff/default/

File: tests/unit/cli/config_test.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+from tests import unittest
+
+from fig.cli import config
+from fig.cli.errors import UserError
+
+import yaml
+import os
+
+
+TEST_YML_DICT_WITH_BACKOFFS = yaml.load("""db:
```
Comment (thaJeztah) [inline tests/unit/cli/config_test.py]: Perhaps change this (and line 26) `_BACKOFFS` -> `_DEFAULTS` to be consistent.

File: tests/unit/cli/config_test.py [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    def test_fully_interpolated_errors(self):
+        if 'TEST_VALUE' in os.environ:
+            del os.environ['TEST_VALUE']
+        self.assertRaises(UserError, config.resolve_environment_vars, "${TEST_VALUE}")
+
+    def test_functional_backoffs_as_dict(self):
+        d = config.with_environment_vars(TEST_YML_DICT_WITH_BACKOFFS)
+
+        # tests the basic structure and functionality of backoffs
+        self.assertEqual(d['web']['build'], '.')
```
Comment (thaJeztah) [inline tests/unit/cli/config_test.py]: s/backoff/default/

File: tests/unit/cli/config_test.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+        # test that environment variables with backoffs are handled in lists
```
Comment (thaJeztah) [inline tests/unit/cli/config_test.py]: s/backoff/default/

File: tests/unit/cli/config_test.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        self.assertEqual(d['web']['ports'][0], '"8000:8000"')
+
+        # test that environment variables with backoffs are handled in dictionaries
```
Comment (thaJeztah) [inline tests/unit/cli/config_test.py]: s/backoff/default/

... And a couple other ones below

**Comment by aanand** [comment]
Thanks!

1. I'm similarly wary of regular expressions, but if there's no practical alternative then fine. In either case we're going to need an *extensive* set of unit tests to cover edge cases and such.
2. How do I escape the dollar-curly-brace syntax, if I actually want the string `${FOO}` to appear?
3. A single, squashed commit is preferred.

**Comment by thaJeztah** [comment]
Good points by Aanand.

I added a couple of nits while glancing over the PR.

Thanks very much, this looks like a very useful feature that I'll definitely use!

**Comment by relwell** [comment]
@aanand I've added in support to escape the text as "\${FOO}". Please refer to (the following test)[https://github.com/docker/fig/pull/845/files#diff-a1643645d7f79dbd5684ed0631ab1066R50] and let me know if you think anything is missing here.

File: fig/cli/config.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+                     for subkey, subvalue in value.items()])
+    elif type(value) == list:
+        return [resolve_environment_vars(x) for x in value]
+    else:
+        return resolve_environment_vars(value)
+
+
+def from_yaml_with_environment_vars(yaml_filename):
+    """
+    Resolves enviornment variables in values defined by a YAML file and transformed into a dict
```
Comment (thaJeztah) [inline fig/cli/config.py]: Typo; enviornment

File: tests/unit/cli/config_test.py [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        self.assertEqual(expected, config.resolve_environment_vars("${TEST_VALUE}"))
+
+        # if we have a match with a default value
+        self.assertEqual(expected, config.resolve_environment_vars("${TEST_VALUE:localhost}"))
+
+        # escaping should prevent interpolation
+        escaped_no_backoff = "\${TEST_VALUE}"
+        escaped_with_backoff = "\${TEST_VALUE:localhost}"
+        self.assertEqual(escaped_no_backoff, escaped_no_backoff)
+        self.assertEqual(escaped_with_backoff, escaped_with_backoff)
```
Comment (thaJeztah) [inline tests/unit/cli/config_test.py]: Feels like I'm bullying here (sorry); s/backoff/default/

File: tests/unit/cli/config_test.py [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+  image: postgres
+web:
+  build: ${DJANGO_BUILD:.}
+  command: python manage.py runserver 0.0.0.0:8000
+  volumes:
+    - .:/code
+  environment:
+    DJANGO_ENV: ${TEST_VALUE:production}
+  ports:
+    - ${DJANGO_PORT:"8000:8000"}
```
Comment (thaJeztah) [inline tests/unit/cli/config_test.py]: Interested to see a test case and/or instructions for providing port numbers <= 60 via environment variables, to test for this: https://github.com/docker/fig/blob/master/docs/yml.md#ports. Don't know if it's possible to do, but thinking of:

    - ${PORT_A}:${PORT_B}

How should that be done to prevent fig from interpreting them incorrectly?

File: tests/unit/cli/config_test.py (hunk 13 above)
Comment (relwell) [inline tests/unit/cli/config_test.py]: My bad -- thought I got them all :)

**Comment by relwell** [comment]
I've added some basic regex tokenization support that will allow us to provide multiple environment variables per line, and accompanying tests.

**Comment by relwell** [comment]
@aanand , it's been 22 days since #846 was opened, and not much has been decided. The community seems to be heavily in favor of delivering a solution to #495 sooner rather than later. 

If you take #426 into account, there has been an open community request for a feature like this for over 120 days.

Since this seems to be close to the direction we want to take, and there have been no additional revision requests, I'd appreciate a merge on this so that we can show some progress on this issue. If you can't merge now, I think we'd all appreciate some kind of timeline or feedback, since we haven't seen much activity from the admins on this issue for the last couple of weeks.

**Comment by relwell** [comment]
FYI: I've updated the PR to support the new docker compose format.

**Comment by relwell** [comment]
@aanand @thaJeztah @dnephin would it be possible to get some status on this PR, and any outstanding issues that may need to be addressed to get it merged in?

**Comment by thaJeztah** [comment]
Unfortunately, I'm not a maintainer, so I cannot give you an update. I *like* this feature and certainly hope this gets approved/merged. It may need a final review to check if everything is okay to merge.

**Comment by aanand** [comment]
I'm going to review this soon - we've a lot on our plate with the impending Compose release. I'd like to get this merged soon after that release, but it needs thorough inspection and fuzz-testing.

**Comment by hawknewton** [comment]
For what it's worth I find environment-specific configuration values shouldn't have defaults.

This usually goes one of two ways: either the developer enters development-time defaults and forgets to update them for production (?!) or the defaults are for production and you have to override them in your development/QA environments.  Having your dev/QA deploy jobs diverge from your production deploy jobs is a clear anti-pattern not to mention the terrifying possibility that somebody forgets to override the production database information for a QA deploy.

I'm super-interested in coming up with a mechanism by which a given profile could be activated as defined in the the yaml that would tailor the application for a given environment or class of environment.  When in QA mode I might use a lower TTL to facilitate testing.  When in development mode maybe I force the developer to provide database connection information via environment variables, etc.

Anyway, just my two cents.

**Comment by torbjornvatn** [comment]
Any news on when we can expect this PR to appear in a release. It's a feature I really need to make make docker-compose scheme to work

**Comment by hadim** [comment]
Any news here ?

**Comment by relwell** [comment]
Is there a plan for integrating this pull request, ever?

I'm happy to rebase and re-push, but I'm not going to bother if you guys don't have a plan to integrate these changes.

My ask is that if there is _not_ a plan to integrate these changes, that you reject the CR, and close #495 as a "won't fix" feature request.

**Comment by aanand** [comment]
I've put the current state of my thinking at https://github.com/docker/compose/issues/1377. Summary: this is the best design so far, but the implementation isn't there yet.

**Comment by relwell** [comment]
I'll leave you guys to worry about handling this then. I'm tired of wasting my time with this.
